[Autograd Testing] Test failure in parent graph before child reentrant task
finishes.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35268 [Autograd Testing] Test failure in parent graph before child reentrant task
finishes.**
finishes.**
finishes.**
* #35223 [Autograd Testing] Add a test where child reentrant task fails.
finishes.**
* #35223 [Autograd Testing] Add a test where child reentrant task fails.

finishes.

Context: https://github.com/pytorch/pytorch/issues/34367

Test: Mixed with Errors -> Reentrant on different devices -> Make parent error
before child finishes

Differential Revision: [D20612919](https://our.internmc.facebook.com/intern/diff/D20612919/)

pritamdamania committed:
[Autograd Testing] Test failure in parent graph before child reentrant task
finishes.

Context: https://github.com/pytorch/pytorch/issues/34367

Test: Mixed with Errors -> Reentrant on different devices -> Make parent error
before child finishes

Differential Revision: [D20612919](https://our.internmc.facebook.com/intern/diff/D20612919/)

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
nit: You don't need to save the whole list here right? Just the last value?

albanD commented on test/test_autograd.py:
The second condition is to avoid waiting forever here right?
Maybe raise an error when we reach it stating that we leaked something?

albanD commented on test/test_autograd.py:
nit: you want to sleep before reading the memory no?

albanD commented on test/test_autograd.py:
What is the reason for the two different branches in the parent graph? I think it's fine if a CPU Node starts a GPU-only reentrant workload. Or did you wanted to explicitly avoid this case?

albanD commented on test/test_autograd.py:
nit: in general avoid empty forwards in Function. It is a special case that can lead to weird behavior when mixed with inplace. I usually do `return inp.clone()`.

pritamdamania87 commented on test/test_autograd.py:
Thanks for the suggestion, although which inplace operation are you referring to here?

pritamdamania87 commented on test/test_autograd.py:
Don't we need two branches in some sense to ensure the CPU thread has some work when we start a GPU task? If there is only a single branch of execution, wouldn't the CPU thread anyways have to wait for the GPU task to finish (since the next CPU task depends on the output of it)?

pritamdamania87 commented on test/test_autograd.py:
When the unit test finishes the test framework already does this validation here: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_utils.py#L538. So if we exceed that timeout without cleaning up all the memory, that check would fail and the unit test would fail as well.

pritamdamania committed:
Update on "[Autograd Testing] Test failure in parent graph before child reentrant task
finishes."

finishes.**
* #35223 [Autograd Testing] Add a test where child reentrant task fails.

finishes.

Context: https://github.com/pytorch/pytorch/issues/34367

Test: Mixed with Errors -> Reentrant on different devices -> Make parent error
before child finishes

Differential Revision: [D20612919](https://our.internmc.facebook.com/intern/diff/D20612919/)

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
Ho ok, but you could raise a more informative error here like "Is the child graph stuck"?
If you don't do that, it would be nice to at least have a comment saying that the test infra is going to raise an error when we leak here.

albanD commented on test/test_autograd.py:
There is not here, it's a nit. So it is fine to do it here. It is more of a general guideline when writing custom Functions.

albanD commented on test/test_autograd.py:
Yes, my bad.

albanD reviewed (approved):
Looks good, a small comment for people not familiar with the overall testing infra would be nice.

pritamdamania committed:
Update on "[Autograd Testing] Test failure in parent graph before child reentrant task
finishes."

finishes.**
finishes.**
* #35223 [Autograd Testing] Add a test where child reentrant task fails.
finishes.**
* #35223 [Autograd Testing] Add a test where child reentrant task fails.

finishes.

Context: https://github.com/pytorch/pytorch/issues/34367

Test: Mixed with Errors -> Reentrant on different devices -> Make parent error
before child finishes

Differential Revision: [D20612919](https://our.internmc.facebook.com/intern/diff/D20612919/)

[ghstack-poisoned]

wanchaol commented on test/test_autograd.py:
maybe we should combine this together with `SimulateBackwardError`?

wanchaol commented on test/test_autograd.py:
can we reduce the sleep time? this test alone takes about 30s..
